File: navpipe/dataset/nav.py
from __future__ import annotations
import asyncio
import polars as pl
from typing import Iterable
from navpipe.transport.async_http import AsyncNavPipeHTTP
from navpipe.transform.nav import nav_response_to_df
from navpipe.models.nav import NavResponse, DateRange
import msgspec


async def _fetch_nav_df(
    http: AsyncNavPipeHTTP,
    scheme_code: int,
    date_range: DateRange | None,
    sem: asyncio.Semaphore,
) -> pl.DataFrame:
    params: dict[str, str] = {}
    if date_range:
        params["startDate"] = date_range.start_date
        params["endDate"] = date_range.end_date

    async with sem:
        raw_bytes = await http.get_raw(f"/mf/{scheme_code}", params=params)

    # Decode directly into NavResponse
    typed = msgspec.json.decode(raw_bytes, type=NavResponse)

    return nav_response_to_df(typed)


async def fetch_nav_history_bulk(
    scheme_codes: Iterable[int],
    *,
    date_range: DateRange | None = None,
    max_concurrency: int = 5,
    rate_limit_per_sec: int | None = None,
) -> pl.DataFrame:
    sem = asyncio.Semaphore(max_concurrency)

    async with AsyncNavPipeHTTP(rate_limit_per_sec=rate_limit_per_sec) as http:
        tasks = [_fetch_nav_df(http, code, date_range, sem) for code in scheme_codes]

        dfs = await asyncio.gather(*tasks)

    return pl.concat(dfs, how="vertical")

File: src/dataset/nav.rs
use crate::schema::nav::{DateRange, NavResponse};
use crate::transform::nav::nav_response_to_df;
use crate::transport::client::AsyncNavPipeHTTP;
use futures::future::join_all;
use polars::prelude::*;
use std::sync::Arc;
use tokio::sync::Semaphore;

/// Fetch NAV data concurrently for an owned list of scheme codes
pub async fn fetch_nav_history_bulk(
    scheme_codes: Vec<u32>, // Accepts the owned list
    _date_range: Option<DateRange>,
    max_concurrency: usize,
) -> Result<DataFrame, PolarsError> {
    // We initialize the client once to reuse the connection pool
    let client = AsyncNavPipeHTTP::new();
    let sem = Arc::new(Semaphore::new(max_concurrency));

    let tasks: Vec<_> = scheme_codes
        .into_iter() // Consumes the Vec, turning each u32 into an owned value
        .map(|code| {
            let client_clone = client.clone();
            let sem_clone = sem.clone();

            tokio::spawn(async move {
                let _permit = sem_clone.acquire().await.unwrap();

                // Path construction
                let path = format!("mf/{}", code);

                let resp: NavResponse = client_clone
                    .get_path(&path)
                    .await
                    .map_err(|e| PolarsError::ComputeError(e.to_string().into()))?;

                nav_response_to_df(&resp)
            })
        })
        .collect();

    let results = join_all(tasks).await;
    let mut dfs = Vec::with_capacity(results.len());

    for res in results {
        // Handle Task Join Error (Tokio) then the Polars Result
        let df = res.map_err(|e| PolarsError::ComputeError(e.to_string().into()))??;
        dfs.push(df);
    }

    // Combine all individual DataFrames into one large one
    polars::functions::concat_df_diagonal(&dfs)
}

